activerecord/lib/active_record/connection_adapters/abstract/query_comment.rb

```diff
@@ -0,0 +1,143 @@
+# frozen_string_literal: true
+
+require "active_support/core_ext/module/attribute_accessors_per_thread"
+
+module ActiveRecord
+  module ConnectionAdapters
+    # SQL query comments set at the adapter level
+    module QueryComment
+      mattr_accessor :prepend_comment, default: false
+      def add_query_comment?
+        return false unless ActiveRecord::Base.query_comment_enabled
+        return false if Context.components.empty?
+        true
+      end
+
+      def add_query_comment_to_sql(sql)
+        return sql unless add_query_comment?
+        Context.set(adapter: self)
+        comment = Context.comment
+        if comment.present? && !sql.include?(comment)
+          sql = if prepend_comment
+            "#{comment} #{sql}"
+          else
+            "#{sql} #{comment}"
+          end
+        end
+        sql
+      end
+
+      # Maintains a user-defined context for all queries and constructs an SQL comment by
+      # calling methods listed in :components
+      #
+      # Additional information can be pushed to the context to be referenced by methods
+      # defined in framework or application initializers.
+      module Context
+
+        mattr_accessor :components, :lines_to_ignore
+        mattr_accessor :cache_query_comment, default: false
+        thread_mattr_accessor :context, default: {}
+        thread_mattr_accessor :cached_comment
```

Comment:
If these are only accessible by `Context.components` etc then `instance_accessor: false` should be set. IE if we don't want Rails or the apps to have access to `@@components`.

Reply:
Ah yea - `components` & `lines_to_ignore` are the only ones which have (documented) direct access in the original gem, but the others can change 👍

Reply:
scratch that - was thinking about that incorrectly - they can all change :)